File: leetcode-algorithms/016. 3Sum Closest/solution.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <numeric>
using namespace std;

int threeSumClosest(vector<int>& nums, int target) {
	if(nums.size()<=3)
		return accumulate(nums.begin(),nums.end(),0);
	sort(nums.begin(), nums.end());

	int ans = nums[0]+nums[1]+nums[2];
	int n = nums.size();

	for (int i = 0; i < n-2; ++i)
	{
		int j=i+1,k=n-1;
		while(j<k)
		{
			int sum = nums[i]+nums[j]+nums[k];
			if(abs(target-ans)>abs(target-sum))
			{
				ans = sum;
				if(ans==target)
					return ans;
			}
			(sum>target)?k--:j++;
		}
	}
	return ans;

}

int main()
{
	std::vector<int> v{-1,2,1,4};
	int res = threeSumClosest(v,1);
	std::cout<<res<<endl;
	return 0;
}

File: leetcode-algorithms/016. 3Sum Closest/solution.py
def threeSumClosest(nums, target):
	"""
	:type nums: List[int]
	:type target: int
	:rtype: int
	"""
	nums.sort()

	if len(nums)<=3:
		return sum(nums)

	n = len(nums)
	res = 0x7fffffff
	for i in range(0,n-2):
		j = i + 1
		k = n - 1
		while j<k:
			s = nums[i]+nums[j]+nums[k]
			if abs(target-res)>abs(target-s):
				res = s
				if res == target:
					return res
			if s > target:
				k -= 1
			else:
				j += 1
	return res

if __name__ == '__main__':
	nums = [-1,2,1,-4]
	target = 1
	res = threeSumClosest(nums,target)
	print(res)